runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -10283,49 +10262,108 @@ genHeapAlloc(TR::Node * node, TR::Instruction *& iCursor, bool isVariableLen, TR
+            // Zero the remainder.
+            if(isLargeConstantLen)
+               {
+               // The remainder is a constant value between 1 and 256 for large constant length objects and displacement is 0.
+               allocSize = ((allocSize - 1) % 256) + 1;
+               switch (allocSize)
+                  {
+                  case 1:
+                     iCursor = generateSIInstruction(cg, TR::InstOpCode::MVI, node, generateS390MemoryReference(addressReg, 0, cg), 0, iCursor);
+                     break;
+                  case 2:
+                     iCursor = generateSILInstruction(cg, TR::InstOpCode::MVHHI, node, generateS390MemoryReference(addressReg, 0, cg), 0, iCursor);
+                     break;
+                  case 4:
+                     iCursor = generateSILInstruction(cg, TR::InstOpCode::MVHI, node, generateS390MemoryReference(addressReg, 0, cg), 0, iCursor);
+                     break;
+                  case 8:
+                     iCursor = generateSILInstruction(cg, TR::InstOpCode::MVGHI, node, generateS390MemoryReference(addressReg, 0, cg), 0, iCursor);
+                     break;
+                  default:
+                     iCursor = generateSS1Instruction(cg, TR::InstOpCode::XC, node, (allocSize - 1), generateS390MemoryReference(addressReg, 0, cg), generateS390MemoryReference(addressReg, 0, cg), iCursor);
+                     break;
+                  }
+               }
+            else
+               {
+               TR::LabelSymbol *exrlTargetLabel = generateLabelSymbol(cg);
+               iCursor = generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, fillerRemLabel);
+               iCursor = generateS390LabelInstruction(cg, TR::InstOpCode::label, node, exrlTargetLabel);
+               iCursor = generateSS1Instruction(cg, TR::InstOpCode::XC, node, 0, generateS390MemoryReference(addressReg, 0, cg), generateS390MemoryReference(addressReg, 0, cg), iCursor);
+               iCursor = generateS390LabelInstruction(cg, TR::InstOpCode::label, node, fillerRemLabel);
+               iCursor = generateRILInstruction(cg, TR::InstOpCode::EXRL, node, tmp, exrlTargetLabel);
+               }
+            srm->reclaimScratchRegister(tmp);
+            srm->reclaimScratchRegister(addressReg);
+            }
+         else
+            {
+            int32_t displacement = 0;
+            while (allocSize > 256)
+               {
+               iCursor = generateSS1Instruction(cg, TR::InstOpCode::XC, node, 255, generateS390MemoryReference(resReg, displacement, cg), generateS390MemoryReference(resReg, displacement, cg), iCursor);
+               displacement += 256;
+               allocSize -=256;
+               }
+            switch (allocSize)
+               {
+               case 1:
+                  iCursor = generateSIInstruction(cg, TR::InstOpCode::MVI, node, generateS390MemoryReference(resReg, displacement, cg), 0, iCursor);
```

Comment:
Some of this code is getting repeated for same function, does it make sense to refactor that to common function and call that?